jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,335 @@
+
+                        /*
+                         * only return the cycle if it is negative (in the first iteration it can be non-negative)
+                         */
+                        if(cycleResult.getCost() < 0) {
+                            return cycleResult;
+                        }
+                    }
+
+                    for(E e : graph.outgoingEdgesOf(tail)) {
+                        V currentVertex = graph.getEdgeTarget(e);
+                        if(!path.getLabels().contains(labels.get(currentVertex)) && path.getCost() + graph.getEdgeWeight(e) < 0) { // extend the path if the extension is still negative a correctly labeled
+                            LabeledPath<V> newPath = path.clone();
+                            newPath.addVertex(currentVertex, graph.getEdgeWeight(e), labels.get(currentVertex));
+                            pathsLengthKplus1.add(newPath);
+
+                            /*
+                             * check if paths are dominated, i.e. if the path is definitely worse than other paths
+                             * and does not have to be considered in the future
+                             */
+                            testDomination(path, pathsLengthKplus1);
+                        }
+                    }
+                }
+            }
+            // update k and the corresponding sets
+            k += 1;
+            pathsLengthK = pathsLengthKplus1;
+            pathsLengthKplus1 = new LinkedHashSet<>();
+        }
+
+        return new LabeledPath<>();
+    }
+
+    /**
+     * removes all paths that are dominated from all calculated paths of length k + 1.
+     * This is important out of efficiency reasons, otherwise many unnecessary paths may
+     * be considered in further calculations.
+     *
+     *
+     * @param path the currently calculated path
+     * @param PathsLengthKplus1 all before calculated paths of length k + 1
+     */
+    private void testDomination(LabeledPath<V> path, Set<LabeledPath<V>> PathsLengthKplus1) {
+        boolean removePath = false;
+
+        for(Iterator<LabeledPath<V>> it = PathsLengthKplus1.iterator(); it.hasNext();) {
+            LabeledPath<V> pathInSetKplus1 = it.next();
+
+            if(path == pathInSetKplus1) {
+                continue;
+            }
+
+            if(pathInSetKplus1.dominates(path)) {
+                /*
+                 * we have to delete path after the for loop otherwise we get an ConcurrentModificationException
+                 */
+                removePath = true;
+                /*
+                 * we can break because domination of paths is transitive, i.e. pathInSetKplus1 already removed the dominated paths
+                 */
+                break;
+            }
+            if(path.dominates(pathInSetKplus1)) {
+                PathsLengthKplus1.remove(pathInSetKplus1);
```

Comment:
Why is this not at risk of ConcurrentModificationException as well?  Are you somehow relying on the LinkedHashSet ordering?  That's not a good idea.

This is a case where it.remove can be used instead (since pathInSetKplus1 was the last return value of it.next).